File: packages/ios/Sources/Components/DynamicIsland/DynamicIsland.swift
import SwiftUI

/// Abstract shape bridging the native ActivityKit Dynamic Island constraints.
public struct KozmosDynamicIsland<Expanded: View, CompactLeading: View, CompactTrailing: View, Minimal: View>: View {
    public enum IslandState {
        case compact
        case expanded
        case minimal
    }
    
    public let state: IslandState
    public let expandedContent: () -> Expanded
    public let compactLeading: () -> CompactLeading
    public let compactTrailing: () -> CompactTrailing
    public let minimalContent: () -> Minimal
    
    public init(
        state: IslandState = .compact,
        @ViewBuilder expandedContent: @escaping () -> Expanded,
        @ViewBuilder compactLeading: @escaping () -> CompactLeading,
        @ViewBuilder compactTrailing: @escaping () -> CompactTrailing,
        @ViewBuilder minimalContent: @escaping () -> Minimal
    ) {
        self.state = state
        self.expandedContent = expandedContent
        self.compactLeading = compactLeading
        self.compactTrailing = compactTrailing
        self.minimalContent = minimalContent
    }
    
    public var body: some View {
        ZStack {
            // Expanded
            if state == .expanded {
                expandedContent()
                    .padding()
                    .frame(width: 360, height: 160)
                    .background(Color.black)
                    .clipShape(RoundedRectangle(cornerRadius: 32, style: .continuous))
                    .transition(.scale.combined(with: .opacity))
            }
            
            // Compact Mode: a 240 × 44 capsule, as React and Compose draw it.
            // It was 36 high and as wide as its container until 2026-09-22.
            if state == .compact {
                HStack(spacing: 8) {
                    compactLeading()
                    Spacer()
                    compactTrailing()
                }
                .padding(.horizontal, 16)
                .frame(width: 240, height: 44)
                .background(Color.black)
                .clipShape(Capsule())
                .transition(.scale.combined(with: .opacity))
            }
            
            // Minimal Mode: a 56 circle, as React and Compose draw it (48
            // until 2026-09-22).
            if state == .minimal {
                minimalContent()
                    .frame(width: 56, height: 56)
                    .background(Color.black)
                    .clipShape(Circle())
                    .transition(.scale.combined(with: .opacity))
            }
        }
        .animation(.spring(response: 0.3, dampingFraction: 0.7, blendDuration: 0), value: state)
        .preferredColorScheme(.dark)
    }
}

File: packages/android/src/main/java/com/kozmos/components/DynamicIsland/DynamicIsland.kt
package com.kozmos.components.dynamicisland

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens
import com.kozmos.tokens.LocalKozmosUseDarkTokens

enum class KozmosDynamicIslandState {
    Compact,
    Expanded,
    Minimal
}

@Composable
fun KozmosDynamicIsland(
    modifier: Modifier = Modifier,
    state: KozmosDynamicIslandState = KozmosDynamicIslandState.Compact,
    compactLeading: @Composable () -> Unit = {},
    compactTrailing: @Composable () -> Unit = {},
    expandedContent: @Composable () -> Unit = {},
    minimalContent: @Composable () -> Unit = {}
) {
    val width by animateDpAsState(
        targetValue = when (state) {
            KozmosDynamicIslandState.Expanded -> 360.dp
            KozmosDynamicIslandState.Minimal -> 56.dp
            KozmosDynamicIslandState.Compact -> 240.dp
        },
        label = "KozmosDynamicIslandWidth"
    )
    val height by animateDpAsState(
        targetValue = when (state) {
            KozmosDynamicIslandState.Expanded -> 160.dp
            KozmosDynamicIslandState.Minimal -> 56.dp
            KozmosDynamicIslandState.Compact -> 44.dp
        },
        label = "KozmosDynamicIslandHeight"
    )
    val radius by animateFloatAsState(
        targetValue = if (state == KozmosDynamicIslandState.Expanded) 32f else 100f,
        label = "KozmosDynamicIslandRadius"
    )

    // The island is black in both themes, so what sits in it reads from the
    // dark palette, as SwiftUI scopes its island to the dark scheme. Until
    // 2026-09-22 its content took the page's theme: dark ink on black in light
    // mode.
    CompositionLocalProvider(LocalKozmosUseDarkTokens provides true) {
        Surface(
            modifier = modifier.width(width).height(height),
            shape = RoundedCornerShape(radius.dp),
            color = Color.Black,
            contentColor = KozmosThemeTokens.primitivesColorsForeground100,
            tonalElevation = 8.dp,
            shadowElevation = 16.dp
        ) {
            when (state) {
                // 16 in from the pill's ends, and 16 around the expanded
                // card, as React and SwiftUI inset them. Compose drew both
                // flush to the edge until 2026-09-22.
                KozmosDynamicIslandState.Compact -> {
                    Row(
                        modifier = Modifier.padding(horizontal = 16.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Box(modifier = Modifier.weight(1f), contentAlignment = Alignment.CenterStart) {
                            compactLeading()
                        }
                        Spacer(modifier = Modifier.width(8.dp))
                        Box(modifier = Modifier.weight(1f), contentAlignment = Alignment.CenterEnd) {
                            compactTrailing()
                        }
                    }
                }
                KozmosDynamicIslandState.Expanded -> {
                    Box(modifier = Modifier.padding(16.dp), contentAlignment = Alignment.Center) {
                        expandedContent()
                    }
                }
                KozmosDynamicIslandState.Minimal -> {
                    Box(contentAlignment = Alignment.Center) {
                        minimalContent()
                    }
                }
            }
        }
    }
}
